nameko_sqlalchemy/transaction_retry.py

```diff
@@ -1,28 +1,61 @@
+
+import eventlet
 import functools
 import operator
 
 import wrapt
 from sqlalchemy import exc
 
 
-def transaction_retry(wrapped=None, session=None):
+def transaction_retry(wrapped=None, session=None, retries=1, backoff_factor=0):
 
     if wrapped is None:
-        return functools.partial(transaction_retry, session=session)
+        return functools.partial(
+            transaction_retry, session=session,
+            retries=retries, backoff_factor=backoff_factor)
 
     @wrapt.decorator
     def wrapper(wrapped, instance, args, kwargs):
 
-        try:
-            return wrapped(*args, **kwargs)
-        except exc.OperationalError as exception:
-            if exception.connection_invalidated:
-                if isinstance(session, operator.attrgetter):
-                    session(instance).rollback()
-                elif session:
-                    session.rollback()
+        @retry(retries, backoff_factor, exc.OperationalError)
+        def run_or_rollback():
+            try:
                 return wrapped(*args, **kwargs)
-            else:
+            except exc.OperationalError as exception:
+                if exception.connection_invalidated:
+                    if isinstance(session, operator.attrgetter):
+                        session(instance).rollback()
+                    elif session:
+                        session.rollback()
                 raise
 
+        return run_or_rollback()
+
     return wrapper(wrapped)  # pylint: disable=E1120
+
+
+def retry(retries, backoff_factor, exceptions):
```

Comment:
Nameko has a retry decorator utility too: https://github.com/nameko/nameko/blob/master/nameko/utils/retry.py

The backoff algorithm is different, but you could use that instead?

Reply:
Yes, definitely.  One note on mulling this over: Using the "normal" retry decorator there is no way to have an immediate retry other than setting the "delay" parameter to 0, in which case there is no backoff at all (just keep retrying immediately indefinitely).  In the use case here of database connection errors, the immediate retry is important in the pure "reconnection" case, which is probably 95% of cases, so it kind of makes sense to start off with an immediate retry in this case.  However, I don't think this is worth having two different behaviors in the retries.

Reply:
No problem, as you saw, I was having the same debate in my head.

I reverted back but added a bit of documentation.  In addition, the api and algorithm is now the same as used in the urllib3/requests retry algorithm requests: total, backoff_factor, backoff_max (https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#urllib3.util.retry.Retry) which should be a good base for expected behavior.